plugin/core/windows.py

```diff
@@ -522,12 +522,27 @@ class WindowRegistry:
     def __init__(self) -> None:
         self._enabled = False
         self._windows: dict[int, WindowManager] = {}
-        client_configs.set_listener(self._on_client_config_updated)
+        client_configs.set_listeners(self._on_client_config_updated, self._on_userprefs_updated)
 
     def _on_client_config_updated(self, config_name: str | None = None) -> None:
         for wm in self._windows.values():
             wm.get_config_manager().update(config_name)
 
+    def _on_userprefs_updated(self) -> None:
+        sublime.set_timeout_async(self._on_userprefs_updated_async)
+
+    def _on_userprefs_updated_async(self) -> None:
+        for wm in self._windows.values():
+            wm.on_diagnostics_updated()
+            for session in wm.get_sessions():
+                session.redraw_config_status_async()
+                for sb in session.session_buffers_async():
+                    sb.redraw_document_links_async()
+                    if not userprefs().semantic_highlighting:
+                        sb.clear_semantic_tokens_async()
+                for sv in session.session_views_async():
+                    sv.redraw_diagnostics_async()
```

Comment:
I wonder if it would be better to delegate all those actions to Session/SessionBuffer/SessionView. We'd trigger `on_settings_changed` on `Session` and it would propagate this to SessionBuffers->SessionViews by calling public `on_settings_changed` on those.

We wouldn't need to expose public methods for each feature separately and I feel like the logic be contained better, without having to delegate responsibility to trigger correct actions to other places.

Reply:
https://github.com/sublimelsp/LSP/pull/2448/commits/32d3bef74016697caeeb5d5ff6a905814425203b

I named it `on_userprefs_changed_async`